### PR TITLE
Endpoints

### PR DESCRIPTION
**Pull request by GreyMatteOr**
# Description

Added a post endpoint to the server to start and run the game.
It branches into 3 different helper functions:

 - `startGame` (should maybe rename to setWord) takes a word to guess and sets it in the game
 - `joinGame` allows players to join, as well as to designate if they should generate the word (at the moment, there are no safeguards in place, so a person may overwrite and 'steal' the generator role)
- `makeGuess` allows players to guess letters and words. Updates the game, and return display information.

There are some preliminary safeguards and error handling hooks set up, but we can flesh those out later.

# How was this tested

In POSTMAN. Can demo over slack call later.

# Thoughts
Still need to write endpoints that let players leave, but I think that will be better to do with `socket.on('disconnect')` which will be utilized later.

So far this can interact with the page without sockets, but it will be MUCH more responsive and better with `sockets`. I think next step is to hook that up.

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -13,6 +15,19 @@ io.on('connection', (socket) => {
   })
 })
 
+app.post('/', (req, resp) => {
```
Comment (holladayian) [inline server.js]: what is `'/'` in reference to this line?

File: server.js [hunk 2]
```diff
@@ -13,6 +15,19 @@ io.on('connection', (socket) => {
...
+  if (req.act === 'join') {
+    joinGame(resp);
+  }
+  if (req.act === 'word') {
+    startGame(req, resp)
+  }
+  if (req.act === 'guess') {
+    makeGuess(req, resp)
+  }
+  resp.status(400).json("Incorrect 'act' verb");
```
Comment (holladayian) [inline server.js]: Good lookin, although I don't think a user can hit this intentionally anyways (without changing the functionality with dev tools). Still good to have!

File: server.js (hunk 1 above)
Comment (GreyMatteOr) [reply server.js]: The is the equivalent to the home-page. It's the 'ending', or everything that follows the `url`.

File: server.js (hunk 2 above)
Comment (GreyMatteOr) [reply server.js]: True! I also included it for development purposes, too! Now we can see with more accuracy where things might go wrong.

File: server.js [hunk 3]
```diff
@@ -2,38 +2,55 @@ const app = require('express')();
...
 })
 
 app.get('/', (req, resp) => {
   resp.send('<h3>Connected</h3>');
 })
 
+joinGame(isGenerator, resp) {
+  let id = Date.now();
+  while (players.includes(id)) {
+    id -= 1;
```
Comment (holladayian) [inline server.js:39]: why the -= 1?

File: server.js (hunk 1 above)
Comment (holladayian) [reply server.js]: Gotcha. So we wouldn't need to update it for things that are getting sent to and from /gamepage? I suppose the Gamepage component would send that action back up to App anyways

File: server.js [hunk 4]
```diff
@@ -39,21 +39,48 @@ joinGame(isGenerator, resp) {
...
-
+makeGuess({ guess, id }) {
+  if (!guess) {
+    return resp.status(400).json(`Guess is missing. Guess: ${guess}.`);
+  }
+  if (!game.verifyGen(id)) {
+    return resp.status(401).json(`Generator is not allowed to guess. ID provided: ${id}.`);
+  }
+  if (!players.includes(id)) {
+    return resp.status(401).json(`Only current players may guess. ID provided: ${id}.`);
```
Comment (holladayian) [inline server.js:78]: Yeah, get the heck outta here, non players!

File: server.js [hunk 5]
```diff
@@ -1,24 +1,102 @@
-const app = require('express')();
```
Comment (Chriscastanuela) [inline server.js:1]: I like this one better but that's just my preferential thing

File: server.js (hunk 1 above)
Comment (GreyMatteOr) [reply server.js]: Sort of. The actual `fetch` has it's url set already, and you can call that from any compenent at any path, and it will always access the same `url`.

TL;DR: client-url !== server-url

File: server.js [hunk 6]
```diff
@@ -39,21 +39,48 @@ joinGame(isGenerator, resp) {
...
+  });
 }
 
-
+makeGuess({ guess, id }) {
+  if (!guess) {
+    return resp.status(400).json(`Guess is missing. Guess: ${guess}.`);
+  }
+  if (!game.verifyGen(id)) {
+    return resp.status(401).json(`Generator is not allowed to guess. ID provided: ${id}.`);
```
Comment (holladayian) [inline server.js:75]: So I think this will be part of the conditional rendering from React. If a player's state is set to isGen = true, they will not see an input field to make a guess, and thus will not have access to it's functionality.

File: server.js [hunk 7]
```diff
@@ -1,24 +1,102 @@
-const app = require('express')();
+const express = require('express');
+const app = express();
 const server = require('http').createServer(app);
 const io = require('socket.io')(server);
+const Game = require('./Game/Game.js');
+let game = new Game, players = [], generator;
```
Comment (Chriscastanuela) [inline server.js:6]: Are you sure we don't need parens after `new Game`?

File: server.js (hunk 3 above)
Comment (GreyMatteOr) [reply server.js:39]: To guarantee unique ID. Since multiple people could in theory try to access the game at the exact same millisecond, just added an easy way to keep them unique.

File: server.js (hunk 1 above)
Comment (holladayian) [reply server.js]: gotcha

File: server.js (hunk 4 above)
Comment (GreyMatteOr) [reply server.js:78]: LMAO

File: server.js (hunk 6 above)
Comment (GreyMatteOr) [reply server.js:75]: Awesome. I have no idea about client stuff, but when we link it up, we can remove redundancy later.

I do like the idea of making the game-host `hack-proof` though.... XD

**Review by Chriscastanuela — APPROVED** [review]

File: server.js (hunk 5 above)
Comment (GreyMatteOr) [reply server.js:1]: I needed access to the express library to import `use(express.json())` a few lines down, unfortunately. I also loved the style way more!
